File: scripts/eval/phase2_competition_eval.py
"""Phase 2.3 Competition Evaluation."""

import argparse
import json
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))

from longmem_dmr_smoke import eval_cargo_profile, repo_root


def run_phase2_competition_eval(dataset_dir: Path, output_path: Path, tag: str) -> dict:
    cargo_profile = eval_cargo_profile()
    cmd = ["cargo", "run"]
    if cargo_profile == "release":
        cmd.append("--release")
    cmd.extend(
        [
            "-p",
            "synapse-eval",
            "--bin",
            "kr-phase2-competition-eval",
            "--",
            "--dataset-dir",
            str(dataset_dir),
            "--json",
            str(output_path),
            "--tag",
            tag,
        ]
    )
    result = subprocess.run(cmd, cwd=repo_root(), text=True, capture_output=True)
    if result.returncode != 0:
        raise RuntimeError(
            f"kr-phase2-competition-eval failed\nSTDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}"
        )
    print(result.stdout, end="")
    if result.stderr:
        print(result.stderr, end="", file=sys.stderr)
    report = json.loads(output_path.read_text(encoding="utf-8"))
    report["cargo_profile"] = cargo_profile
    return report


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Run Phase 2.3 memory competition evaluation."
    )
    parser.add_argument(
        "--dataset-dir",
        type=Path,
        default=repo_root() / "crates/eval/datasets/cognitive_memory",
    )
    parser.add_argument(
        "--output-path",
        type=Path,
        default=repo_root() / "crates/eval/reports/phase2-competition-eval.json",
    )
    parser.add_argument("--tag", default="phase2-competition-eval")
    args = parser.parse_args()

    args.output_path.parent.mkdir(parents=True, exist_ok=True)
    report = run_phase2_competition_eval(args.dataset_dir, args.output_path, args.tag)
    print(
        "summary: cases={} baseline={:.4f} competition={:.4f} decision_mismatch_delta={:+} causal_order_error_delta={:+} suppression={:.4f} influence_shift={:.4f} pass={}".format(
            report["case_count"],
            report["baseline"]["score"],
            report["competition"]["score"],
            report["delta"]["decision_mismatch"],
            report["delta"]["causal_order_error"],
            report["delta"]["suppression_correctness"],
            report["delta"]["influence_shift"],
            report["pass"],
        )
    )
    print(f"Saved to {args.output_path}")
    if report["pass"] and report["case_count"] == 200:
        print("Phase 2.3 Competition Evaluation Completed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/eval/src/bin/phase2_competition_eval.rs
use anyhow::{Context, Result};
use clap::Parser;
use std::path::PathBuf;
use synapse_eval::Phase2CompetitionEvaluator;

#[derive(Parser)]
#[command(
    name = "kr-phase2-competition-eval",
    about = "Synapse Phase 2.3 memory competition evaluation over cognitive memory benchmark"
)]
struct Cli {
    /// Directory containing cognitive memory benchmark TOML suites.
    #[arg(long, default_value = "crates/eval/datasets/cognitive_memory")]
    dataset_dir: PathBuf,
    /// Write the Phase2CompetitionEvaluationReport JSON to this path.
    #[arg(long)]
    json: PathBuf,
    /// Tag to embed in the report.
    #[arg(long, default_value = "phase2-competition-eval")]
    tag: String,
}

fn main() -> Result<()> {
    let cli = Cli::parse();
    let report = Phase2CompetitionEvaluator::evaluate(&cli.dataset_dir, cli.tag)?;

    println!("=== Synapse Phase 2.3 competition evaluation ===");
    println!("tag:                         {}", report.tag);
    println!("dataset_dir:                 {}", report.dataset_dir);
    println!("cases:                       {}", report.case_count);
    println!(
        "baseline:                    {} ({:.4})",
        report.baseline.name, report.baseline.score
    );
    println!(
        "competition:                 {} ({:.4})",
        report.competition.name, report.competition.score
    );
    println!(
        "decision_mismatch_delta:     {:+}",
        report.delta.decision_mismatch
    );
    println!(
        "causal_order_error_delta:    {:+}",
        report.delta.causal_order_error
    );
    println!(
        "suppression_correctness:     {:.4}",
        report.delta.suppression_correctness
    );
    println!(
        "influence_shift:             {:.4}",
        report.delta.influence_shift
    );
    println!("pass:                        {}", report.pass);

    if let Some(parent) = cli.json.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&cli.json, serde_json::to_string_pretty(&report)?)
        .with_context(|| format!("writing JSON to {}", cli.json.display()))?;
    eprintln!(
        "wrote Phase 2.3 competition report to {}",
        cli.json.display()
    );
    Ok(())
}
